### ISSUE TITLE
memory leak with scipy 0.16 spatial cKDEtree 

### ISSUE THREAD

**Issue by evanmason** [scipy/scipy#5630]
I believe there is some kind of memory leak in scipy.spatial.cKDEtree v0.16.  The script below demonstrates the behaviour I see; with scipy 0.15 memory usage is constant, while with scipy 0.16 it increases until out of memory.  I see this behaviour on a MacBook and two Linux distributions (Mint and Mageia 5).
```
# %run test_ckdtree.py
'''
If run with scipy 0.15 increases in memory usage tail
off after a few tens of iterations.
With scipy 0.16 memory usage increases continuously.

Tested on an Intel MacBook (Python 2.7.11), Linux Mint
and Linux Mageia
'''
import numpy as np
from scipy.spatial import cKDTree

import resource

FILLVAL = 99.

# Make some data
dx, dy = 0.05, 0.05
y, x = np.mgrid[slice(1, 5 + dy, dy),
                slice(1, 5 + dx, dx)]
z = np.sin(x)**10 + np.cos(10 + y*x) * np.cos(x)
z_copy = np.empty_like(z)

for i in xrange(500000):

    z_copy[:] = z

    # Place FILLVAL in z_copy at random number of random locations
    mask = np.random.random_integers(0,
                                    z.size - 1,
                                    np.random.randint(50) + 5)
    z_copy.flat[mask] = FILLVAL

    igood = np.vstack(np.where(x != FILLVAL)).T
    ibad = np.vstack(np.where(x == FILLVAL)).T

    mem_use = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

    tree = cKDTree(igood)
    dist, iquery = tree.query(ibad, k=4, p=2)

    new_mem_use = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

    if new_mem_use > mem_use:
        mem_use = new_mem_use
        print 'mem increase to %s ' % mem_use

print 'done'
```

**Comment by rgommers** [comment]
By putting parts of this script outside the for-loop, you can see that the ``query`` method is fine, but constructing a new ``cKDTree`` increases memory usage with 264 bytes. Those never get freed it looks like. 

@sturlamolden maybe you have an explanation?

**Comment by sturlamolden** [comment]
I will investigate. Any memory leak is a bug.

**Comment by sturlamolden** [comment]
```__dealloc__``` is misspelled ```__deallocate__``` and never called.

**Comment by sturlamolden** [comment]
Posted fixes for master (backport to 0.17.x) and 0.16.x. Two PRs because the file ckdtree.pyx is moved to a different location, and automatic backport to 0.16.x is impossible.

**Comment by sturlamolden** [comment]
Fixed when #5635 and #5636 are pulled. Unit test proves the problem is gone.

**Comment by ev-br** [comment]
Fixed by gh-5635, closing. Thanks @evanmason @sturlamolden @rgommers
